### ISSUE TITLE
Problem with defgeneric and defmethod

### ISSUE THREAD

**Issue by DamienCassou** [Alhadis/language-emacs-lisp#2]
Hi,

thank you for your excellent work. I've just noticed that `defgeneric`, `cl-defgeneric`, `defmethod` and `cl-defmethod` are not recognized as keywords (at least in Github, I haven't checked Atom):

```emacs
(defgeneric my/entity-name (entity)
  "Return the name of ENTITY.")

(defmethod my/entity-name ((entity person))
  (person-name entity))

(cl-defgeneric my/entity-name (entity)
  "Return the name of ENTITY.")

(cl-defmethod my/entity-name ((entity person))
  (person-name entity))
```

**Comment by Alhadis** [comment]
:wave: Hey @DamienCassou,

Since you're not an Atom user, I'm assuming you found your way here through GitHub's [list of repositories](https://github.com/github/linguist/tree/master/vendor) it uses to provide syntax highlighting across the site. They recently introduced a new grammar compiler which dropped any pattern rules exceeding a predetermined token length. Since the number of functions in Emacs' standard library is colossal, so too is the pattern used to match each one of them in source code.

I fixed this in [`e6024060`](https://github.com/Alhadis/language-emacs-lisp/commit/e6024060ddcec488d29aec39d3b35cb77de3be0b), but the changes won't propagate across the site until there's been a new release of Linguist (which I believe is due to happen in the next few days).

In short: don't worry, everything will be fine. :wink:

**Comment by Alhadis** [comment]
My preferred editor is Atom, actually. :wink: Emacs is my secondary editor, and that huge-ass regex was actually scraped from [`doc.endlessparentheses.com`](http://doc.endlessparentheses.com/). :sweat_smile: What's the sane way of doing that purely with Elisp?

**Comment by DamienCassou** [comment]
Here is a way to get a macro matching all macros and special forms:

```emacs
(progn
  ;; require a few standard libraries to get their definitions
  (require 'cl-lib)
  (require 'subr-x)
  (require 'rx)
  (let ((all-macros (let (cands)
                      (mapatoms
                       (lambda (x)
                         (when (or (macrop x) (special-form-p x))
                           (push (symbol-name x) cands))))
                      cands)))
    (rx-to-string `(or ,@all-macros))))
```

**Comment by DamienCassou** [comment]
You can evaluate the above with `C-x C-e` and get a large string (available in the `*Messages*` buffer).

**Comment by Alhadis** [comment]
Thank you! :grinning: I'll work off this sometime to extract a full list of functions as well as macros.
